File: server/island_list.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Copyright (C) 2026 Yevgeniy Kropochev <y.kropochev87@gmail.com>
# This file is part of WS Messenger. See LICENSE for terms.

"""
The island's signed list: where to reach it, and which keys to trust.

One document serves two features that would otherwise each need their own
trust root:

  * entry points, so a client can fail over between several addresses of this
    island without any of them being substitutable by whoever answers first;
  * relay transport keys, so a relayed direct message is sealed to a key the
    client can authenticate. Whoever substitutes that key reads the metadata of
    every envelope - which is the whole property the relay network exists to
    protect. See docs/internal/message-relay-assessment.md §7.5.

Clients pin the signing key on first contact and refuse anything not signed by
it afterwards. That is trust-on-first-use, with the same weakness as every
TOFU: the first contact must be honest.

## Canonical form

The signature covers

    b"ws-island-list-v1" + utf8(canonical_json(payload))

Domain-separated, so a signature made here can never be replayed as a
signature over something else.

`canonical_json` must produce, byte for byte, what `_stableJson` in both
clients' crypto-utils produces. That holds only if the payload obeys two
rules, which are enforced below rather than trusted:

  * no floats - JSON.stringify(1.0) is "1" while Python would write "1.0";
  * ASCII object keys - JS sorts keys by UTF-16 code unit and Python by code
    point, which differ above the BMP.

String VALUES may be any Unicode: ensure_ascii=False makes Python emit raw
characters exactly as JSON.stringify does.
"""

from __future__ import annotations

import base64
import json
import os
import time
from typing import Any

from fastapi import APIRouter, HTTPException
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey
from cryptography.hazmat.primitives.serialization import Encoding, PublicFormat

SIGN_DOMAIN = b"ws-island-list-v1"


def b64d(s: str) -> bytes:
    s = s.strip().replace("-", "+").replace("_", "/")
    return base64.b64decode(s + "=" * (-len(s) % 4))


def b64e(b: bytes) -> str:
    return base64.b64encode(b).decode("utf-8")


def _assert_canonicalizable(value: Any, path: str = "payload") -> None:
    if isinstance(value, bool) or value is None or isinstance(value, str):
        return
    if isinstance(value, float):
        raise ValueError(f"{path}: floats are not canonicalizable across clients")
    if isinstance(value, int):
        return
    if isinstance(value, list):
        for i, v in enumerate(value):
            _assert_canonicalizable(v, f"{path}[{i}]")
        return
    if isinstance(value, dict):
        for k, v in value.items():
            if not isinstance(k, str) or not k.isascii():
                raise ValueError(f"{path}: object keys must be ASCII, got {k!r}")
            _assert_canonicalizable(v, f"{path}.{k}")
        return
    raise ValueError(f"{path}: unsupported type {type(value).__name__}")


def canonical_json(payload: dict) -> str:
    """Byte-identical to `_stableJson` in both clients, given the rules above."""
    _assert_canonicalizable(payload)
    return json.dumps(payload, sort_keys=True, separators=(",", ":"), ensure_ascii=False)


def signing_message(payload: dict) -> bytes:
    return SIGN_DOMAIN + canonical_json(payload).encode("utf-8")


class IslandListConfig:
    def __init__(self) -> None:
        raw = (os.getenv("ISLAND_SIGNING_KEY_B64") or "").strip()
        self.signing_key: Ed25519PrivateKey | None = (
            Ed25519PrivateKey.from_private_bytes(b64d(raw)) if raw else None
        )
        self.island_id = (os.getenv("ISLAND_ID") or "").strip()
        self.version = int(os.getenv("ISLAND_LIST_VERSION") or "1")
        self.entry_points = json.loads(os.getenv("ISLAND_ENTRY_POINTS") or "[]")
        self.relays = json.loads(os.getenv("ISLAND_RELAYS") or "[]")

    @property
    def enabled(self) -> bool:
        return self.signing_key is not None and bool(self.island_id)

    def public_key_bytes(self) -> bytes:
        assert self.signing_key is not None
        return self.signing_key.public_key().public_bytes(Encoding.Raw, PublicFormat.Raw)


# The document changes only when the operator changes the configuration, but
# `issued_at` moves with the clock. Re-signing on every request would make an
# unauthenticated endpoint do public-key work on demand, which is a cheap way to
# burn someone else's CPU. Sign at most this often; freshness at this
# granularity is far finer than the staleness window clients apply.
LIST_CACHE_S = 60


def build_router(*, transport_keys: list[dict] | None = None) -> APIRouter:
    """
    transport_keys: [{"kid": "<b64>", "public_key_b64": "<b64>"}] - the relay
    ingress keys this island will accept envelopes for. Passed in rather than
    read here, because relay_ingress owns them.
    """
    router = APIRouter()
    cfg = IslandListConfig()
    cached: dict[str, Any] = {"at": 0.0, "doc": None}

    @router.get("/.well-known/wsapp-island")
    async def island_list() -> dict[str, Any]:
        if not cfg.enabled:
            raise HTTPException(status_code=404, detail="island list not configured")

        nowf = time.time()
        if cached["doc"] is not None and nowf - cached["at"] < LIST_CACHE_S:
            return cached["doc"]

        payload: dict[str, Any] = {
            "island_id": cfg.island_id,
            "version": cfg.version,
            # Seconds, not milliseconds, and an int: a float would not
            # canonicalize identically in both languages.
            "issued_at": int(time.time()),
            "entry_points": cfg.entry_points,
            "relays": cfg.relays,
            "transport_keys": transport_keys or [],
        }

        sig = cfg.signing_key.sign(signing_message(payload))
        doc = {
            "payload": payload,
            "sig_b64": b64e(sig),
            # Present for trust-on-first-use only. After the first contact a
            # client MUST verify against the key it pinned and ignore this
            # field entirely - otherwise the signature proves nothing, since
            # an attacker would simply ship their own key alongside it.
            "signing_key_b64": b64e(cfg.public_key_bytes()),
        }
        cached["at"] = nowf
        cached["doc"] = doc
        return doc

    return router

File: Android/src/services/island-list.js
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2026 Yevgeniy Kropochev <y.kropochev87@gmail.com>
// This file is part of WS Messenger. See LICENSE for terms.

// Verification of an island's signed list: its entry points, its relays and
// the transport keys a relayed message may be sealed to.
//
// This is the trust root for two features at once. Without it, failover would
// follow whatever address answered first, and a relayed direct message would be
// sealed to whatever transport key the network handed over — and whoever
// substitutes that key reads the metadata of every envelope, which is the
// entire property the relay network exists to protect.
//
// Trust-on-first-use: the signing key is pinned on first contact and every
// later document must be signed by it. TOFU's weakness is TOFU's weakness — the
// first contact has to be honest — but from then on an attacker needs the key,
// not merely a position on the network.
//
// This file is duplicated byte-for-byte in two places:
//   chrome_extension/island-list.js
//   Android/src/services/island-list.js
// Keep them identical; island-list-parity.test.js enforces it. That is why it
// takes no platform imports: Ed25519 verification, canonical JSON and base64
// decoding are all injected, because the two clients implement them on
// different crypto stacks.

// `root` is passed in rather than read from `self`/`globalThis` inside: the two
// clients lint under different environments, and a bare reference to either
// name is an error in one of them.
(function (root) {
  "use strict";

  // Domain separation: a signature made over an island list can never be
  // replayed as a signature over anything else.
  const SIGN_DOMAIN = "ws-island-list-v1";

  const MAX_ENTRY_POINTS = 8;
  const MAX_RELAYS = 16;
  const MAX_TRANSPORT_KEYS = 4;

  // A document older than this is stale, but staleness is NOT fatal: a client
  // cut off from every entry point cannot refresh, and refusing to use the last
  // known list would turn a censorship event into a total outage. Callers are
  // told, and decide.
  const STALE_AFTER_MS = 30 * 24 * 60 * 60 * 1000;

  function isPlainObject(v) {
    return !!v && typeof v === "object" && !Array.isArray(v);
  }

  function str(v) {
    return typeof v === "string" ? v.trim() : "";
  }

  /**
   * An address the client may actually dial.
   *
   * The document is signed, so a bad value here means the island is attacking
   * its own users - but "signed" is not "safe to hand to fetch()". Only http
   * and https get through, which keeps javascript:, file: and data: out of a
   * list that a caller might use without looking.
   */
  function safeUrl(raw, schemes) {
    const v = str(raw).replace(/\/+$/, "");
    if (!v) return "";
    let u;
    try { u = new URL(v); } catch (e) { return ""; }
    if (schemes.indexOf(u.protocol) < 0) return "";
    return u.origin;
  }

  /**
   * Canonical JSON — must be byte-identical to what the island signed.
   *
   * Constraints the island enforces on its side and this mirrors: no floats
   * (JSON.stringify(1.0) is "1", Python would write "1.0") and ASCII object
   * keys (JS sorts keys by UTF-16 code unit, Python by code point). String
   * values may be any Unicode.
   */
  function canonicalJson(value) {
    if (value === null || typeof value !== "object") return JSON.stringify(value);
    if (Array.isArray(value)) return "[" + value.map(canonicalJson).join(",") + "]";
    const keys = Object.keys(value).sort();
    const parts = [];
    for (let i = 0; i < keys.length; i++) {
      const v = value[keys[i]];
      if (v === undefined) continue;
      parts.push(JSON.stringify(keys[i]) + ":" + canonicalJson(v));
    }
    return "{" + parts.join(",") + "}";
  }

  function signingMessage(payload, utf8Encode) {
    return utf8Encode(SIGN_DOMAIN + canonicalJson(payload));
  }

  /**
   * createIslandListVerifier({ ed25519Verify, b64decode, utf8Encode, now })
   *
   *   ed25519Verify(pubBytes, sigBytes, msgBytes) -> Promise<boolean> | boolean
   *   b64decode(s) -> Uint8Array        (tolerant of base64url and no padding)
   *   utf8Encode(s) -> Uint8Array
   *   now() -> ms
   */
  function createIslandListVerifier(deps) {
    const d = deps || {};
    const ed25519Verify = d.ed25519Verify;
    const b64decode = d.b64decode;
    const utf8Encode = d.utf8Encode;
    const now = d.now || function () { return Date.now(); };
    if (!ed25519Verify || !b64decode || !utf8Encode) {
      throw new Error("createIslandListVerifier: ed25519Verify, b64decode and utf8Encode are required");
    }

    /**
     * @param {object} doc      what the island served
     * @param {object} pin      { signingKeyB64, islandId, version } or null on first contact
     * @returns {Promise<{ok, reason?, payload?, pin?, stale?}>}
     */
    async function verify(doc, pin) {
      if (!isPlainObject(doc) || !isPlainObject(doc.payload)) {
        return { ok: false, reason: "malformed" };
      }
      const payload = doc.payload;

      const islandId = str(payload.island_id);
      const version = Number(payload.version);
      if (!islandId) return { ok: false, reason: "malformed" };
      if (!isFinite(version) || version < 1 || Math.floor(version) !== version) {
        return { ok: false, reason: "malformed" };
      }

      // Which key must have signed this. On first contact the document carries
      // its own key and we are trusting the channel; afterwards the embedded
      // key is ignored entirely, because an attacker would just ship theirs.
      const firstContact = !pin || !str(pin.signingKeyB64);
      const keyB64 = firstContact ? str(doc.signing_key_b64) : str(pin.signingKeyB64);
      if (!keyB64) return { ok: false, reason: "no signing key" };

      if (!firstContact) {
        if (str(pin.islandId) && str(pin.islandId) !== islandId) {
          return { ok: false, reason: "different island" };
        }
        // Rollback protection: a replayed older document could reinstate a
        // retired relay or drop an entry point that was added since.
        const pinnedVersion = Number(pin.version || 0);
        if (isFinite(pinnedVersion) && version < pinnedVersion) {
          return { ok: false, reason: "rollback" };
        }
      }

      let keyBytes;
      let sigBytes;
      try {
        keyBytes = b64decode(keyB64);
        sigBytes = b64decode(str(doc.sig_b64));
      } catch (e) {
        return { ok: false, reason: "malformed" };
      }
      if (!keyBytes || keyBytes.length !== 32) return { ok: false, reason: "malformed" };
      if (!sigBytes || sigBytes.length !== 64) return { ok: false, reason: "bad signature" };

      let good = false;
      try {
        good = await ed25519Verify(keyBytes, sigBytes, signingMessage(payload, utf8Encode));
      } catch (e) {
        good = false;
      }
      if (!good) return { ok: false, reason: "bad signature" };

      const normalized = normalizePayload(payload);
      if (!normalized.entryPoints.length) return { ok: false, reason: "no entry points" };

      const issuedAtMs = Number(payload.issued_at || 0) * 1000;
      const stale = issuedAtMs > 0 && now() - issuedAtMs > STALE_AFTER_MS;

      return {
        ok: true,
        stale: stale,
        payload: normalized,
        pin: { signingKeyB64: keyB64, islandId: islandId, version: version },
      };
    }

    return { verify: verify };
  }

  /** Shape and bound the document so callers never handle raw server input. */
  function normalizePayload(payload) {
    const entryPoints = [];
    const seenEp = {};
    const rawEp = Array.isArray(payload.entry_points) ? payload.entry_points : [];
    for (let i = 0; i < rawEp.length && entryPoints.length < MAX_ENTRY_POINTS; i++) {
      const e = rawEp[i];
      if (!isPlainObject(e)) continue;
      const apiBase = safeUrl(e.apiBase, ["http:", "https:"]);
      if (!apiBase || seenEp[apiBase]) continue;
      seenEp[apiBase] = true;
      entryPoints.push({
        apiBase: apiBase,
        wsBase: safeUrl(e.wsBase, ["ws:", "wss:"]),
        label: str(e.label).slice(0, 40),
      });
    }

    const relays = [];
    const seenRelay = {};
    const rawRelays = Array.isArray(payload.relays) ? payload.relays : [];
    for (let i = 0; i < rawRelays.length && relays.length < MAX_RELAYS; i++) {
      const r = rawRelays[i];
      if (!isPlainObject(r)) continue;
      const id = str(r.id);
      const url = safeUrl(r.url, ["http:", "https:"]);
      if (!id || !url || seenRelay[id]) continue;
      seenRelay[id] = true;
      relays.push({ id: id, url: url, label: str(r.label).slice(0, 40) });
    }

    const transportKeys = [];
    const rawKeys = Array.isArray(payload.transport_keys) ? payload.transport_keys : [];
    for (let i = 0; i < rawKeys.length && transportKeys.length < MAX_TRANSPORT_KEYS; i++) {
      const k = rawKeys[i];
      if (!isPlainObject(k)) continue;
      const kid = str(k.kid);
      const pub = str(k.public_key_b64);
      if (!kid || !pub) continue;
      transportKeys.push({ kid: kid, publicKeyB64: pub });
    }

    return {
      islandId: str(payload.island_id),
      version: Number(payload.version),
      issuedAt: Number(payload.issued_at || 0),
      entryPoints: entryPoints,
      relays: relays,
      transportKeys: transportKeys,
    };
  }

  const API = {
    SIGN_DOMAIN: SIGN_DOMAIN,
    STALE_AFTER_MS: STALE_AFTER_MS,
    MAX_ENTRY_POINTS: MAX_ENTRY_POINTS,
    MAX_RELAYS: MAX_RELAYS,
    canonicalJson: canonicalJson,
    signingMessage: signingMessage,
    normalizePayload: normalizePayload,
    createIslandListVerifier: createIslandListVerifier,
  };

  if (typeof module !== "undefined" && module.exports) module.exports = API;
  else if (root) root.WSIslandList = API;
})(this);
